vllm/v1/worker/tpu_model_runner.py

```diff
@@ -744,12 +910,20 @@ def capture_model(self) -> None:
 
         start = time.perf_counter()
         num_tokens = 16
+        num_reqs_to_sample = MIN_NUM_SEQS
         while True:
-            self._dummy_run(self.kv_caches, num_tokens)
-            logger.info("  -- num_tokens: %d", num_tokens)
-            xm.mark_step()
-            xm.wait_device_ops()
-            if num_tokens >= self.max_num_tokens:
+            # Compile for different sampler outputs in
+            # [MIN_NUM_SEQS, pad(max_num_reqs)]
+            while True:
+                self._dummy_run(self.kv_caches, num_tokens)
+                logger.info("  -- num_tokens: %d, num_seqs: %d", num_tokens,
+                            num_reqs_to_sample)
+                xm.mark_step()
+                xm.wait_device_ops()
+                if num_reqs_to_sample >= self.max_num_reqs:
+                    break
+                num_reqs_to_sample *= 2
+            if num_tokens >= self.scheduler_config.max_num_batched_tokens:
                 break
             num_tokens *= 2
```

Comment:
This doesn't seem right.. each time we are only passing in `num_tokens`, so what is the point of `num_reqs_to_sample` in this loop? Looking around it seems like `num_tokens_to_sample` is missing to be passed in.

This potential explosion of compilation makes me feel a better solution would be to completely separate the graphs for model forward and sampling. This way we can just define a "max sampling batch size" that can be separated - so we end up with M+N compilations rather than M*N

Reply:
I don't think I can do it in in M+N here as sampling will be dependent on both `num_reqs_to_sample` and *`hidden_state`*, so for each different hidden_state shape we still need to compile. But we will only have MN compilations of the sampling step which should be much lighter.
 
I did consider breaking the graph multiple times but iirc the main concern was the interaction with torch.compile.